rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,265 @@
+# RFC 6330 - 2021-04-07 - Converting one log event into multiple log events
+
+This RFC describes an approach for transforming one input log event into multiple log events.
+
+## Scope
+
+In:
+
+- Turning a log event into multiple log events
+
+Out:
+
+- Turning a log event into multiple metrics
+- Turning a metric into multiple metrics
+- Turning a metric into multiple logs
+
+## Motivation
+
+Users have asked [1][] [2][] to be able to transform an incoming event into multiple events. This is useful, for example, when parsing an incoming JSON payload that contains an array for which you'd like to publish an event for each element.
+
+Currently, users are restricted to using the Lua or WASM transform for this, [introducing a substantial bottleneck](https://user-images.githubusercontent.com/316880/105531520-d3643200-5cbf-11eb-8b35-fe1c99e5c254.png).
+
+## Internal Proposal
+
+The proposal is to:
+
+1. Extend `remap` so that if `.` is an array at the end, it will emit one event for each element in that array
+2. Add an `unnest` VRL function that will transform an object into an array of objects using a specified field on the input object
+3. Add an `unnest` transform that can be used for simple cases (could just "expand" into a `remap` transform)
+
+Example input:
+
+```json
+{ "host": "localhost", "events": [{ "message": "foo" }, { "message": "bar" }] }
+```
+
+Remap transform config:
+
+```toml
+[transforms.remap]
+type = "remap"
+source = """
+. = unnest(., "events")
+"""
+```
+
+Ouput:
+
+```json
+{ "host": "localhost", "message": "foo" }
+{ "host": "localhost", "message": "bar" }
+```
+
+This would also be available as an `unnest` transform for simple cases where the user does not need to do any additional mapping:
+
+```toml
+[transforms.unnest]
+type = "unnest"
+field = "events"
+```
+
+Which would have the same effect. I think the implementation could just take advantage of the "expand" feature of transforms that lets them replace themselves at vector start time with another transform.
+
+Additionally, we will provide `only_fields` and `except_fields` as options on the `unnest` function and transform to allow users to select which fields will be kept. These match similar semantics to the `encoding` options on sinks.
```

Comment:
Is there a big need for this, compared to using `del` before calling `unnest`? I guess these arguments could be considered an easier solution to achieve the same, but I'm not convinced it's worth the added complexity.

Reply:
Yeah, it's not a big deal, users could workaround this via calling `del` beforehand. It's just a shortcut.

Though, we don't have an easy way to say "delete all fields _but_ these" yet.